### PR TITLE
fix #10731 ; `runnableExamples "-b:cpp --run:off": code` works

### LINKED ISSUE #10731: runnableExamples that are only compiled
runnableExamples are run by `nim doc`. Some code examples should not be run due to side effects or other reasons, but parsing and compiling the code is still beneficial to prevent code rot.

For example: httpclient fetching a remote URL; deleting files; and so on.

This can become quite dangerous when running `nim doc` on 3rd party libraries where the authors might have written examples without realizing they would be execute automatically!

### PR DESCRIPTION
**Pull request by timotheecour**
* fixes https://github.com/nim-lang/Nim/issues/10731 with a general fix that allows passing arbitrary cmd to a given runnableExamples, as an argument `rdoccmd` to runnableExamples (bringing feature parity with the rst code block nim)

## example
```nim
proc fun*() =
  ## some comment
  runnableExamples: discard # same as before
  runnableExamples "-d:foo -b:cpp":
    import std/compilesettings
    doAssert querySetting(backend) == "cpp"
  runnableExamples "-r:off": ## this one is only compiled
     import std/browsers
     openDefaultBrowser "https://forum.nim-lang.org/"
```

## more examples
see tests in tests/nimdoc/trunnableexamples.nim

## features
* plays well with `--doccmd:cmd`: `--doccmd:cmd` (if passed) is appended (not prepended) to rdoccmd so that we can override a test via cmdline; it'd be easy to add a prepend option if ever needed.

* runnableExamples are grouped by unique `rdoccmd` so that we maintain the "megatest" speed advantages introduced in https://github.com/nim-lang/Nim/commit/5fc78d86b7c3e53c3404816e96790e191c19af8f

* fixes https://github.com/nim-lang/Nim/issues/10731 as a special case: simply write:
```nim
runnableExamples "-r:off": ## this one is only compiled
   import std/browsers
   openDefaultBrowser "https://forum.nim-lang.org/"
```

## workaround `if false:` was not good enough
it's uglier, eg:
```nim
import std/browsers
if false:
  openDefaultBrowser "https://forum.nim-lang.org/"
```

and doesn't work in many cases as a replacement for `compile but not run`:
* eg if some import would cause module level code to run which would fail (bc of some required dependency)
* or if it needs a dll that may not be available at runtime, which is a common problem, eg:
```nim
proc fun*() =
  # this would work fine with `runnableExamples: "-r:off"` and not require `if false:` hack
  runnableExamples:
    import sdl2, opengl
    if false: # this will still give a runtime failure in case sdl2 dll is not available at runtime
      discard sdl2.init(INIT_EVERYTHING)
      var window = createWindow("test", 100, 100, 640, 480, SDL_WINDOW_OPENGL or SDL_WINDOW_RESIZABLE)
      var context = window.glCreateContext()
      loadExtensions()
      var evt = sdl2.defaultEvent
      while true:
        while pollEvent(evt): break
        glClear(GL_COLOR_BUFFER_BIT or GL_DEPTH_BUFFER_BIT)
        delay 1_000
      destroy window
```

* furthermore it didn't handle at all the other use cases enabled by having custom command, eg `-d:cpp --threads`, or `--compileOnly`, etc. 

bottom line: this makes it simple to have self contained examples in a single file, even if individual examples need different cmdline options

### CODE REVIEW

File: compiler/docgen.nim [hunk 1]
```diff
@@ -26,6 +26,16 @@ const
...
+  Example = ref object
+    ## a single runnableExamples
+    rdoccmd: string
+  ExampleGroup = ref object
+    ## a group of runnableExamples with same rdoccmd
+    rdoccmd: string ## from 1st arg in `runnableExamples(rdoccmd): body`
+    docCmd: string ## from user config, eg --doccmd:-d:foo
+    code: string ## contains imports; each import contains `body`
+    index: int ## group index
+    examples: seq[Example]
```
Comment (Clyybber) [inline compiler/docgen.nim]:
```suggestion
    examples: seq[string]
```
and remove the `Example` type

Comment (timotheecour) [reply compiler/docgen.nim]: done to unblock this PR, but I preferred the way I wrote it as it's more self documenting and I intended to build upon `Example` in future PR; `compiler/docgen.nim` deals with `runnableExamples` so it's natural for it to have a type representing a runnableExample

**Comment by timotheecour** [comment]
@araq friendly ping on this (so I can avoid some workarounds in another PR)

**Comment by timotheecour** [comment]
thanks!
